### PR TITLE
feat: Add support for hints

### PR DESCRIPTION
**Pull request by zawan-ila**
### Description:

This pull request adds support for hints. 

### Details:
If the user starts the game with the `--hint-mode` flag, he/she can type `hint` at the terminal to get a hint. The hint will appraise the user of the synonyms or antonyms of the hidden word. At most two hints can be availed. It should be noted however that sometimes only one or no hints may be provided

### To Do:

- [ ] Add tests

### CODE REVIEW

File: wordle.py [hunk 1]
```diff
@@ -1,6 +1,37 @@
 import random
+import sys
+from nltk.corpus import wordnet as wn
```
Comment (azanbinzahid) [inline wordle.py]: see if we can get a simpler offline dictionary rather this dependency. Log your approach and decision in PR desc.

File: wordle.py [hunk 2]
```diff
@@ -1,6 +1,37 @@
...
+def remove_duplicates_and_truncate(*lists):
+    '''
+    Takes in a number of lists, removes duplicates from each of them,
+    truncates each of them to the first three elements, and returns a list
+    of all the resulting lists
+    '''
+    filtered_lists = []
+    for lis in lists:
+        lis = list(dict.fromkeys(lis))[:3]
+        filtered_lists.append(lis)
```
Comment (azanbinzahid) [inline wordle.py]: read more on map/reduce in python and try if you can achieve the same using them

File: wordle.py [hunk 3]
```diff
@@ -27,6 +58,8 @@ def invalidate(guess):
     if it does not have proper length or has non-alphabetic
     chars
```
Comment (azanbinzahid) [inline wordle.py]: update this with logic-updated and return types.

File: wordle.py [hunk 4]
```diff
@@ -74,11 +107,46 @@ def play(word):
...
+            return 1
+    print("SORRY YOU LOST.")
 
 
 if __name__ == '__main__':
-    random_word = get_random_word("words.txt")
-    print(random_word)
-    play(random_word)
+    print("Starting...")
+    CHOSEN_WORD = get_random_word("words.txt")
```
Comment (azanbinzahid) [inline wordle.py]: decide at this stage whether we have enough information for hints using a lightweight dictionary.

File: wordle.py [hunk 5]
```diff
@@ -74,11 +107,46 @@ def play(word):
...
+
+    if len(sys.argv) > 1:
+        if len(sys.argv[1]) == 5:
+            CHOSEN_WORD = sys.argv[1]
+        if sys.argv[1] == "--hint-mode":
+            HINT_MODE = True
```
Comment (azanbinzahid) [inline wordle.py]: read more on `switch` statements in python.

File: wordle.py [hunk 6]
```diff
@@ -74,11 +107,46 @@ def play(word):
...
+
+    synonyms, antonyms, definitions, examples = ([] for i in range(4))
```
Comment (azanbinzahid) [inline wordle.py]: can you do same with `[] * 4`?

**Review by azanbinzahid — CHANGES_REQUESTED** [review]
Question: why words have been removed from `words.txt`?
Suggestion: use a limited list of words for hint mode that can have all required data e.g. apple, think

File: wordle.py (hunk 1 above)
Comment (zawan-ila) [reply wordle.py]: logged.

File: wordle.py (hunk 4 above)
Comment (zawan-ila) [reply wordle.py]: We have updated the word list to include only those words for which we have hints.

File: wordle.py (hunk 6 above)
Comment (zawan-ila) [reply wordle.py]: Nope. This simply returns an empty list. Even If we do [[]]*4 we get [[], [], [], []] but all the 4 lists have the same underlying reference. Thus any changes in one of them result in changes to all of them.

File: wordle.py (hunk 2 above)
Comment (zawan-ila) [reply wordle.py]: Done. But we think not using map makes it more readable ?

File: wordle.py (hunk 3 above)
Comment (zawan-ila) [reply wordle.py]: Updated

**Comment by azanbinzahid** [comment]
Please extract required data from nltk and remove its dependency in requirements.

File: dictionary.txt [hunk 7]
```diff
@@ -0,0 +1,433 @@
+{"word": "abaft", "synonyms": ["aft", "astern"], "antonyms": ["fore"], "definitions": ["at or near or toward the stern of a ship or tail of an airplane"], "examples": ["ships with square sails sail fairly efficiently with the wind abaft"]}
```
Comment (azanbinzahid) [inline dictionary.txt]: rename this to jsonl if that is appropriate.

File: wordle.py [hunk 8]
```diff
@@ -74,11 +98,27 @@ def play(word):
...
-    print(random_word)
-    play(random_word)
+    print("Starting...")
+    CHOSEN_WORD_DETAILS = get_random_word_details("dictionary.txt")
+
+    if len(sys.argv) > 1 and sys.argv[1] == "--hint-mode":
+        HINT_MODE = True
+
+    # Note that order is important here
+    word, synonyms, antonyms, definitions, examples = \
```
Comment (azanbinzahid) [inline wordle.py]: call get random func directly here

File: wordle.py [hunk 9]
```diff
@@ -74,11 +98,27 @@ def play(word):
 
         if guess.lower() == word:
             print('WOW YOU WON')
-            return
-    print("SORRY YOU LOST. The WORD was", word)
+            return 1
```
Comment (azanbinzahid) [inline wordle.py]: where are we using this value?

File: wordle.py [hunk 10]
```diff
@@ -48,7 +71,7 @@ def take_guess():
     while invalidate(guess):
         cursor_up_line_clear(2 if bad_input else 1)
         bad_input = True
-        print(invalidate(guess))
+        print(invalidate(guess) if invalidate(guess) != 'hint' else get_hint())
```
Comment (azanbinzahid) [inline wordle.py]: limit `invalidate` call to once.

**Review by azanbinzahid — CHANGES_REQUESTED** [review]

File: dictionary.txt (hunk 7 above)
Comment (zawan-ila) [reply dictionary.txt]: Done

File: wordle.py (hunk 9 above)
Comment (zawan-ila) [reply wordle.py]: Nowhere! Replaced with bare `return`

File: wordle.py (hunk 10 above)
Comment (zawan-ila) [reply wordle.py]: Limited to twice since we feel it is more readable that way

File: wordle.py (hunk 8 above)
Comment (zawan-ila) [reply wordle.py]: Done

**Review by azanbinzahid — APPROVED** [review]

**Pull request merged**